### PR TITLE
Enable usage of custom keywords

### PR DESCRIPTION
**Pull request by pedrohjordao**
Work towards closing #17

### CODE REVIEW

**Comment by sassman** [comment]
Hey @pedrohjordao,

this looks really good so far. Thanks for your hard work. I think within the next days I'm able to finish the review and testing.
I would also like to overhaul / enrich the documentation, so maybe a feature branch would help to keep others contribution to it.

File: src/lib.rs [hunk 1]
```diff
@@ -41,6 +45,48 @@ use structopt::StructOpt;
 ///
 /// - `authors`: Author names, taken from usual environment variables (i.e.
 ///   those which are also used by Cargo and git)
+///
+/// The template author can define their own placeholders in their
+/// `cargo-generate.toml` file. This looks like the following:
+///
+/// [placeholders]
```
Comment (sassman) [inline src/lib.rs:52]: this embedded code needs to be wrapped inside a `code` block that starts with `` ```toml ``

you can verify that it works via `cargo doc --verbose` then you see the warnings.

File: src/lib.rs [hunk 2]
```diff
@@ -41,6 +45,48 @@ use structopt::StructOpt;
...
+/// my-placeholder = { type = "string", prompt = "Hello?", choices = ["hello", "world"], default = "hello", regex = "*" }
+/// use-serde = { type = "bool", prompt = "Add serde support?", default = false }
+///
+/// The user of the template will then be asked the the question in "prompt", and must accept the
+/// default value (if provided) or enter a custom value which will be checked against "choices" (if
+/// provided) and regex (if provided).
+///
+/// The placeholder map supports the following keys:
+///
+/// `type` [required]: Must be "string" or "bool"
```
Comment (sassman) [inline src/lib.rs:62]: those brackets in `[required]` hints cargo doc to create a link to a field of a struct, but here we don't reference a field from a struct so it's required to either mask the brackets or remove / replace them with something else.

File: src/lib.rs (hunk 1 above)
Comment (pedrohjordao) [reply src/lib.rs:52]: Yeah, right

The issue is how it looks like in the generated help message, but I'll come up with something

File: src/lib.rs (hunk 2 above)
Comment (pedrohjordao) [reply src/lib.rs:62]: thanks, I'll fix it

File: src/lib.rs (hunk 1 above)
Comment (sassman) [reply src/lib.rs:54]: I suggest we move it all to a module level comment `//!` then it appears on the docs quite nicely.

Comment (pedrohjordao) [reply src/lib.rs:54]: Right, but then structop isn't able to pick it up as the help message, right? Am I missing something?

Comment (sassman) [reply src/lib.rs:54]: I'm not super familiar with `structops` but I was under the impression that the help messages comes directly from the comments above the individual fields. But I can double-check it.

File: src/projectvariables.rs [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+    String(String),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum SupportedVarType {
+    Bool,
+    String,
+}
+
+const INVALID_NAMES: [&str; 2] = ["project-name", "crate_name"];
```
Comment (sassman) [inline src/projectvariables.rs:80]: Maybe the other standard variables should be named here too? Like: `os-arch` and `authors`

File: src/projectvariables.rs [hunk 4]
```diff
@@ -0,0 +1,835 @@
+use liquid_core::Value;
```
Comment (sassman) [inline src/projectvariables.rs:1]: I would suggest renaming this file to have a `_` in the name, like `project_variables.rs`

**Review by sassman — COMMENTED** [review]
Found only few nits, over all very good work! Thanks again.

File: src/lib.rs (hunk 1 above)
Comment (pedrohjordao) [reply src/lib.rs:54]: AFAIK, it gets the help message from the struct documentation itself for the general `-h` message

You can see it if you just do `cargo run` with no extra parameters

File: src/projectvariables.rs (hunk 3 above)
Comment (pedrohjordao) [reply src/projectvariables.rs:80]: makes sense to me

**Comment by sassman** [comment]
Could you also please rebase?

**Comment by pedrohjordao** [comment]
Should I squash the changes also?

**Comment by sassman** [comment]
That would be great, if possible into one commit with a nice commit message that sums it up.

File: src/lib.rs [hunk 5]
```diff
@@ -113,6 +196,20 @@ fn create_git(args: Args, name: &ProjectName) -> Result<()> {
...
 
+fn get_config_file_values(path: &Path) -> Result<HashMap<String, toml::Value>> {
+    match fs::read_to_string(path) {
+        Ok(ref contents) => toml::from_str::<ConfigValues>(contents)
+            .map(|v| v.values)
+            .map_err(|e| e.into()),
+        Err(e) => anyhow::bail!(
+            "{} {} {}",
+            emoji::ERROR,
+            style("VAlues File Error:").bold().red(),
```
Comment (sassman) [inline src/lib.rs:207]: here is a little typo, `VAlues` but should be `Value File Error` possibly

**Review by sassman — CHANGES_REQUESTED** [review]
sorry found one more little thing. But then it looks good.

File: src/lib.rs (hunk 5 above)
Comment (pedrohjordao) [reply src/lib.rs]: Done

**Review by sassman — APPROVED** [review]
Nice job

**Comment by sassman** [comment]
could you please do a rebase on master?

**Pull request merged**